src/numeric/mod.rs

```diff
@@ -1 +1,159 @@
+use crate::imp_prelude::*;
+use num_complex::Complex;
+use rawpointer::PointerExt;
+use std::mem;
+use std::ptr::NonNull;
+
 mod impl_numeric;
+
+impl<T, S, D> ArrayBase<S, D>
+where
+    S: Data<Elem = Complex<T>>,
```

Comment:
Fantastic. I wonder if there should be a constraint on T. We have ensured that Complex<T> is repr(C). Any other constraints on T needed?

The fun return type of the method inspires us to think that what happens if the user inputs an `Array<Complex<ArrayView<_, _>>>` here? ArrayView are Copy but would it truly be a defined operation?

Reply:
> Fantastic. I wonder if there should be a constraint on T. We have ensured that Complex is repr(C). Any other constraints on T needed?

I don't think so, but the guidelines aren't very clear on this. The relevant references here are the [nomicon](https://doc.rust-lang.org/stable/nomicon/other-reprs.html#reprc) and the [unsafe code guidelines](https://rust-lang.github.io/unsafe-code-guidelines/layout/structs-and-tuples.html#c-compatible-layout-repr-c). The `re` field is guaranteed to be at offset `0` in the struct, and the `im` field is guaranteed to be after it, as described by the C17 standard (which is referenced by the unsafe code guidelines):

> Within a structure object, the non-bit-field members and the units in which bit-fields reside have addresses that increase in the order in which they are declared.  A pointer to a structure object,suitably converted, points to its initial member (or if that member is a bit-field, then to the unit in which it resides), and vice versa. There may be unnamed padding within a structure object, but not at its beginning.

The possibility of padding in `Complex<T>` is somewhat unclear. Maybe, to be on the safe side, we should add a few asserts (which should be eliminated by the compiler at compile time):

```rust
assert_eq!(mem::size_of::<Complex<T>>(), mem::size_of::<T>().checked_mul(2).unwrap());
assert_eq!(mem::align_of::<Complex<T>>(), mem::align_of::<T>());
// Non-empty case only:
let ptr = self.ptr.as_ptr();
assert_eq!(
    unsafe { &(*ptr).im as *const T as usize }
        .checked_sub(ptr as usize)
        .unwrap(),
    mem::size_of::<T>(),
);
```

> The fun return type of the method inspires us to think that what happens if the user inputs an `Array<Complex<ArrayView<_, _>>>` here? ArrayView are Copy but would it truly be a defined operation?

Yeah, that would be a weird type. Arrays containing arrays are pretty weird. I don't see any soundness issues with it, though. I think the implementation should work for arbitrary `T`. The return type of `Array::<Complex<ArrayView<'a, _, _>, D>::view_re_im` would be `Complex<ArrayView<'b, ArrayView<'a, _, _>, D>>`. In other words, you'd get a `Complex` struct where the `re` and `im` fields would each be a view containing views.